packages/astro/src/core/endpoint/index.ts

```diff
@@ -128,38 +142,54 @@ export async function callEndpoint<MiddlewareResult = Response | EndpointOutput>
 		response = await renderEndpoint(mod, context, env.ssr);
 	}
 
-	if (response instanceof Response) {
-		attachToResponse(response, context.cookies);
-		return {
-			type: 'response',
-			response,
-		};
-	}
+	// If return simple endpoint object, convert to response
+	if (!(response instanceof Response)) {
+		// In SSR, we convert to a full response with the correct headers to be sent out
+		if (env.ssr && !ctx.route?.prerender) {
+			if (response.hasOwnProperty('headers')) {
+				warn(
+					env.logging,
+					'ssr',
+					'Setting headers is not supported when returning an object. Please return an instance of Response. See https://docs.astro.build/en/core-concepts/endpoints/#server-endpoints-api-routes for more information.'
+				);
+			}
 
-	if (env.ssr && !ctx.route?.prerender) {
-		if (response.hasOwnProperty('headers')) {
-			warn(
-				env.logging,
-				'ssr',
-				'Setting headers is not supported when returning an object. Please return an instance of Response. See https://docs.astro.build/en/core-concepts/endpoints/#server-endpoints-api-routes for more information.'
-			);
+			if (response.encoding) {
+				warn(
+					env.logging,
+					'ssr',
+					'`encoding` is ignored in SSR. To return a charset other than UTF-8, please return an instance of Response. See https://docs.astro.build/en/core-concepts/endpoints/#server-endpoints-api-routes for more information.'
+				);
+			}
 		}
 
-		if (response.encoding) {
-			warn(
-				env.logging,
-				'ssr',
-				'`encoding` is ignored in SSR. To return a charset other than UTF-8, please return an instance of Response. See https://docs.astro.build/en/core-concepts/endpoints/#server-endpoints-api-routes for more information.'
-			);
+		// Passed during SSG where we don't need to return a full response, as we only need
+		// to write the `body` to a file directly.
+		if (returnObjectFormIfAvailable) {
+			return {
+				body: response.body,
+				encoding: response.encoding,
+			};
 		}
+
+		const headers = new Headers();
+
+		const pathname = new URL(ctx.request.url).pathname;
+		const mimeType = mime.getType(pathname) || 'text/plain';
+		headers.set('Content-Type', `${mimeType};charset=utf-8`);
+
+		const bytes = encoder.encode(response.body);
+		headers.set('Content-Length', bytes.byteLength.toString());
+
+		response = new Response(bytes, {
+			status: 200,
+			headers,
+		});
+
+		attachToResponse(response, context.cookies);
```

Comment:
The two places in `vite-plugin-astro-server/route.ts` and `core/app/index.ts` that has this same code is now consolidated here.